## 예외처리
### 프로그램의 오류 에러(Error)
어떤원인에 의해 오동작하거나 비정상적으로 종료되는 경우

### **발생 시점에 따른 구분**
 - 컴파일 에러 (compile-time error)
    → 컴파일 도중 컴파일러가 발견하는 오류 (문법적 오류)

 - 런타임 에러 (runtime error) 
    → 컴파일은 성공적으로 진행되었으나, 프로그램 실행 중에 발생하는 오류

 - 논리적 에러 (logical error) 
    → 컴파일도 되고 프로그램도 실행되었으나, 프로그램이 의도대로 동작하지 않는 오류. 프로그램이 정상적으로 실행되고 종료되었지만 결과가 예상과 다른 경우

### 에러와 예외
*// 프로그램적으로 조치할 수 있으면 예외이고 하드웨어나 외적인 이유면 에러*
#### 에러 (Error) : 발생하면 복구할 수 없는 심각한 오류
 - An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch.
 - 메모리 부족
 - 스택 오버 플로우

#### 예외(Exception) : 프로그래머가 적절한 코드를 통해 대비할 수 있는 오류
 - An *Exception* is an event, witch occurs during the execution of a program, that disrupts the nomal flow of the program's instructions.
 - 클래스 형 변환 실패
 - 파일 읽기 실패

### 예외처리 (Exception Handling)
 - 예외 발생 시 프로그램의 비 정상 종료를 막고 정상적인 실행 상태를 유지하는 것
 - 예외 감지 및 예외 발생 시 동작할 코드 작성 필요

### 예외의 종류 : **체크 예외(Checked Exception)와 언체크 예외(Unchecked Exception)**
 - Checked Exception
    -> 컴파일 시점에서 예외처리가 강제되는 예외   
    ex) IOException, SQLException

 - Unchecked Exception (Runtime Exception)
    -> 실행시 점에서 발생하는 예외로, 컴파일러가 예외처리를 강제하지 않음   
    ex) Nullpoint, ArrayIndexOutOfBound

### 예외 처리 키워드
#### 직접 처리
 - try{} : 예외가 발생할 가능성이 있는 코드를 포함하는 블록
 - catch(예외 매개변수){} : 예외가 발생했을 때 실행할 코드 블록
 - finaly{} :


### 예외 처리 (try ~ catch)
 - 프로그램 실행 시 발생할 수 있는 예외에 대한 대비코드 작성
 - 프로그램의 비정상적인 종료를 막고 정상적인 상태를 유지할 수 있도록 함
 ```java
 try {
     // 예외가 발생할 수 있는 코드
 } catch (SomeException e){
     // 예외가 발생했을 때 처리할 코드
 }
 ```
 - 중괄호(블록) 생략 불가능
 - 중첩 된 예외처리 : 블록 안에 또 다른 try ~ catch 구문이 올 수 있음
 
### 다중 예외처리
 - try 블록에서 여러 종류의 예외가 발생할 경우
 - 하나의 try 블록에 여러 개의 catch 블록 추가 가능
 ```java
  try {
     // 예외가 발생할 수 있는 코드
 } catch (xxException e){
     // 예외가 발생했을 때 처리할 코드
 } catch (YYException e){
     // 예외가 발생했을 때 처리할 코드
 }
 ```
```java
  try {
     // 예외가 발생할 수 있는 코드
 } catch (xxException e | YYException e){
     // 예외가 발생했을 때 처리할 코드
 }
 ```

### 다중 예외처리 유의사항
 - 예외 객체에 해당하는 catch문장을 찾을 때는 **다형성**이 적용 됨
 - 상속관계에서 상위 타입의 예외가 먼저 선언되는 경우 뒤에 등장하는 catch블록은 동작 할 기회가 없음
 - 상속관계에서는 작은 범위에서 큰 범위 순으로 정의

### Exception의 



### 메서드에서 예외 던지기 (throws)
 - 메소드 선언부에 throws 키워드를 사용하여 예외 작성
 - 예외가 여러개일 경우 `,`를 이용하여 나열
 - **메소드가 예외를 처리하는 것이 아닌 전달 (호출한 곳으로)**
 - 조상타입의 예외로 처리 가능 (= 다형성)


### 메소드 제정의(오버라이드)와 throws
메소드 재정의 시 조상클래스의 메소드가 던지는 예외보다 부모 예외를 던질 수 없음
```java
class Parent{
    void methodA() throws IOException{}
    void methodB() throws ClassNotFoundException{}
}
public class Child extends Parent{
    @Override
    void methodA() throws FileNotFoundException{}

    // @Override
    // void methodB() throws Exception{} -> **조상보다 큰 예외를 던질 수 없음**
}
```
오버라이드 시 자식은 부모의 기능을 숨길 수 없음 (접근제한자로 메소드를 감출 수 없음)

### finally
예외 발생 여부와 상관 없이 언제나 실행
```java
try{
    // 예외가 발생할 가능성이 있는 코드
} catch (Exception e){
    // 예외처리코드
} finally{
    // 예외 상관없이 항상 수행해야 하는 코드
}
```
### 자동 자원 반납 구문 (try with resources)
 - try() -> 괄호 안에 객체를 생성하는 코드를 작성하면, 해당 객체는 close()를 호출하지 않아도 블록을 벗어나는 순간 close()가 호출 됨
 - 해당 객체의 클래스가 AutoCloseable이라는 인터페이스를 구현한 클래스여야만 함
 ```java
 try(FileInputStream fis = new FileInputStream("test.txt")){
    // 코드 생략
 } catch (IOException e){
    e.printStackTrace();
 }
 ```

### 사용자 정의 예외
 - 기존에 정의된 예외 이외에 사용자가 직접 정의 예외를 작성할 수 있음
 - 대부분 Exception 또는 RuntimeException 클래스를 상속받아 작성
    - checked exception 활용 : 명시적 예외 처리 또는 throws 필요   
    (코드는 복잡해지지만 처리, 누락 등 오류 발생 가능성은 down)
    - runtime excetpion 활용 : 묵시적 예외 처리 가능   
    (코드가 간결해지지만 예외 처리, 누락 가능성 발생)